iota/transaction/creation.py

```diff
@@ -104,365 +105,367 @@ def increment_legacy_tag(self):
+        """
+        if self.hash:
+            raise RuntimeError('Bundle is already finalized.')
+
+        for addy in inputs:
+            if addy.balance is None:
+                raise with_context(
+                    exc=ValueError(
+                        'Address {address} has null ``balance`` '
+                        '(``exc.context`` has more info).'.format(
+                            address=addy,
+                        ),
+                    ),
+
+                    context={
+                        'address': addy,
+                    },
+                )
+
+            if addy.key_index is None:
+                raise with_context(
+                    exc=ValueError(
+                        'Address {address} has null ``key_index`` '
+                        '(``exc.context`` has more info).'.format(
+                            address=addy,
+                        ),
+                    ),
+
+                    context={
+                        'address': addy,
+                    },
+                )
+
+            self._create_input_transactions(addy)
+
+    def send_unspent_inputs_to(self, address):
+        # type: (Address) -> None
+        """
+        Adds a transaction to send "change" (unspent inputs) to the
+        specified address.
+
+        If the bundle has no unspent inputs, this method does nothing.
+        """
+        if self.hash:
+            raise RuntimeError('Bundle is already finalized.')
+
+        self.change_address = address
+
+    def finalize(self):
+        # type: () -> None
+        """
+        Finalizes the bundle, preparing it to be attached to the Tangle.
+        """
+        if self.hash:
+            raise RuntimeError('Bundle is already finalized.')
+
+        if not self:
+            raise ValueError('Bundle has no transactions.')
+
+        # Quick validation.
+        balance = self.balance
+
+        if balance < 0:
+            if self.change_address:
+                self.add_transaction(ProposedTransaction(
+                    address=self.change_address,
+                    value=-balance,
+                    tag=self.tag,
+                ))
+            else:
+                raise ValueError(
+                    'Bundle has unspent inputs (balance: {balance}); '
+                    'use ``send_unspent_inputs_to`` to create '
+                    'change transaction.'.format(
+                        balance=balance,
+                    ),
+                )
+        elif balance > 0:
+            raise ValueError(
+                'Inputs are insufficient to cover bundle spend '
+                '(balance: {balance}).'.format(
+                    balance=balance,
+                ),
+            )
+
+        # Generate bundle hash.
+        while True:
+            sponge = Kerl()
+            last_index = len(self) - 1
+
+            # type: Tuple[int, ProposedTransaction]
+            for (i, txn) in enumerate(self):
```

Comment:
❔ Does the type hint work when it's on the line preceding the `for` loop?  I thought it had to go on the same line.

Reply:
Example:
Type hint placed on the previous line — note that IDE autocomplete doesn't work:
![type hint on previous line - no autocomplete](https://user-images.githubusercontent.com/202040/35486666-4a4b7134-04d6-11e8-84ca-88584ae7582d.png)

Type hint placed at the end of the line — now IDE autocomplete works as expected.
<img width="543" alt="type hint at end of line - autocomplete works" src="https://user-images.githubusercontent.com/202040/35486684-81b9ea56-04d6-11e8-9b05-41a758db49c1.png">